server/src/main/java/com/google/sps/utils/Ad.java

```diff
@@ -0,0 +1,201 @@
+    public String headline; 
+    public String link;
+    public String content;  
+    public String headlineSentiment; 
+    public String headlineTerms;
+    public String contentSentiment; 
+    public String contentTerms; 
+
+  public Ad() {
+  // Must have a public no-argument constructor for Firesbase
+  }  
+
+  private Ad(AdBuilder builder) {
+    this.id = builder.id;
+    this.advertiser = builder.advertiser; 
+    this.startDate = builder.startDate; 
+    this.endDate = builder.endDate; 
+    this.impressionsMin = builder.impressionsMin;
+    this.impressionsMax = builder.impressionsMax; 
+    this.ageTargetingEnabled = builder.ageTargetingEnabled;
+    this.genderTarget = builder.genderTarget; 
+    this.geoTarget = builder.geoTarget; 
+    this.spendMin = builder.spendMin; 
+    this.spendMax = builder.spendMax; 
+    this.headline = builder.headline; 
+    this.link = builder.link; 
+    this.content = builder.content; 
+    this.headlineSentiment = builder.headlineSentiment;
+    this.headlineTerms = builder.headlineTerms; 
+    this.contentSentiment = builder.contentSentiment; 
+    this.contentTerms = builder.contentTerms; 
+  }
+
+  @Override 
+  public String toString() {
+    String s = new StringBuilder()
+      .append("ID: " + this.id)
+      .append("\nAdvertiser: " + this.advertiser)
+      .append("\nStart Date: " + this.startDate.toString())
+      .append("\nEnd Date: " + this.endDate.toString())
+      .append("\nMinimum Number of Impressions: " + this.impressionsMin)
+      .append("\nMaximum Number of Impressions: " + this.impressionsMax)  
+      .append("\nAge Targeting Enabled: " + this.ageTargetingEnabled)  
+      .append("\nGender Targets: " + this.genderTarget)
+      .append("\nGeo Targets: " + this.geoTarget)
+      .append("\nSpend Min: " + this.spendMin)
+      .append("\nSpend Max: " + this.spendMax) 
+      .append("\nHeadline: " + this.headline) 
+      .append("\nLink: " + this.link)
+      .append("\nContent: " + this.content) 
+      .append("\nHeadline Sentiment: " + this.headlineSentiment) 
+      .append("\nHeadline Terms: " + this.headlineTerms) 
+      .append("\nContent Sentiment: " + this.contentSentiment) 
+      .append("\nContent Terms: " + this.contentTerms) 
+      .toString(); 
+    return s; 
+  }
+
+  public static class AdBuilder {
+
+    public String id; 
+    public String advertiser;  
+    public String startDate;
+    public String endDate; 
+    public long impressionsMin; 
+    public long impressionsMax; 
+    public boolean ageTargetingEnabled; 
+    public List<String> genderTarget; 
+    public List<String> geoTarget;
+    public long spendMin; 
+    public long spendMax; 
+    public String headline; 
+    public String link;
+    public String content;  
+    public String headlineSentiment; 
+    public String headlineTerms;
+    public String contentSentiment; 
+    public String contentTerms; 
+
+    public AdBuilder id(String adId) {
+      this.id = adId; 
```

Comment:
this method's name is not clear as it doesn't reveal the function's intention (set the field name and return the whole object).

=> Change the function's name to something like `setId(String adId)` and have a separate method to return the whole AdBuilder like `AdBuilder.getInstance()` (remember we had something similar for Datastore?)

Reply:
+1